ehrql/dataset_parameters.py

```diff
@@ -0,0 +1,95 @@
+import logging
+import sys
+from argparse import ArgumentParser
+from dataclasses import dataclass
+from typing import Any
+
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class Parameter:
+    name: str
+    type: callable = str
+    default: Any | None = None
+
+
+def get_dataset_parameter(name, type: callable = str, default: Any | None = None):  # NOQA: A002
+    """
+    Retrieve dataset definition parameters that have been passed on the command line.
+
+    Dataset parameters are passed after a `--`, in the format `--key value`
+
+    For example, to pass values for two parameters, `max_age` and `region`:
+
+    ```
+    generate-dataset /path/to/dataset_definiton/.py -- --max_age 70 --sex female
+    ```
+
+    Multiple values can be passed per parameter, separated by a space, and will be parsed
+    as a list, e.g.
+    ```
+    generate-dataset /path/to/dataset_definiton/.py -- --sex male female
+    ```
+
+    For each parameter, call `get_dataset_paramter()`, e.g.:
+    ```
+        sex = get_dataset_parameter(name="sex")
+    ```
+
+    Parameters are parsed as strings, with a default value `None`, unless a `type` and/or
+    default value is explicitly specified. `type` can be any callable that takes a single string argument;
+    e.g. to convert to an integer, and default to 100 if no value is passed:
+    ```
+        max_age = get_dataset_parameter(name="max_age", type=int, default=100)
+    ```
+
+    To use in a dataset definition:
+
+    ```
+    from ehrql import create_dataset, get_dataset_parameter
+
+    max_age = get_dataset_parameter(name="max_age", type=int, default=20)
+    sex = get_dataset_parameter(name="sex", default="male")
+
+    dataset = create_dataset()
+    age = patients.age_on("2025-01-01")
+
+    dataset.define_population((age < max_age) & (patients.sex==sex))
+    ```
+    """
+    # find matching user args
+    target_arg = f"--{name}"
+    user_args = []
+    if target_arg in sys.argv:
+        user_args.append(target_arg)
+        arg_index = sys.argv.index(target_arg)
+        for arg in sys.argv[arg_index + 1 :]:
+            if arg.startswith("--"):
+                break
+            user_args.append(arg)
+
+    parser = ArgumentParser()
+    # We use nargs="+" and action="extend" so we can return either a single value or a list,
+    # depending on what values have been provided
+    parser.add_argument(f"--{name}", type=type, action="extend", nargs="+", default=[])
+
+    # Note that we could use parse_known_args(sys.argv[1:]) which won't raise errors if there are
+    # undefined args. However, parse_known_args does prefix-matching
```

Comment:
I think it might be possible to disable prefix matching by passing `allow_abbrev=False` to `ArgumentParser()`:
https://docs.python.org/3.10/library/argparse.html#prefix-matching

Not actually tried this though!

Reply:
It works!